### ISSUE TITLE
DbSet.Add() has different behaviour to DbContext.Set<T>().Add()

### ISSUE THREAD

**Issue by JackSteel97** [dotnet/efcore#16984]
## What is not working
Please correct me if anything sounds wrong.

This is my understanding of the uses of these two methods:
`DbSet.Add()` is used to add a new item to the database when we know which 'table' we want to add it to at design time. e.g. `DbContext.Accounts.Add(newAccount)`.

`DbContext.Set<T>().Add()` is used when we don't know the type of entity being added until runtime. e.g. `DbContext.Set<T>().Add(newItem)` Where `T` is `Account` at runtime.

We noticed an issue on one of our production sites where duplicate entries were being created when using `DbContext.Set<T>().Add()`. I was only able to reproduce the issue by manually forcing two requests into the controller that executes this add. We're still not sure how two requests could happen in this manner but that issue is separate to this.

### Simplified reproduction set up.
So, we have a controller to add a new entry, this controller calls a generic add method which uses `DbContext.Set<T>().Add()` to actually perform the addition. 

### Issue.
If the controller is hit twice at the same time (simulated for testing) with the same data, the `DbContext.Set<T>().Add()` line runs twice before `DbContext.SaveChanges()` is called at the end of the controller. This creates duplicate rows in the database with everything the same except the auto-incrementing primary key.

However, if for testing purposes I switch the `DbContext.Set<T>().Add()` to the equivalent non-generic version i.e. `DbContext.Accounts.Add(newAccount)`, and then perform the same test - hitting the controller twice with the same data at the same time. This time, only one of the requests succeeds (the first one). The first request correctly inserts the data to the database, but when the second request calls `DbContext.SaveChanges` it throws the following error.

```
Microsoft.EntityFrameworkCore.DbUpdateException: An error occurred while updating the entries. See the inner exception for details. ---> System.Data.SqlClient.SqlException: Cannot insert explicit value for identity column in table 'updated_accounts' when IDENTITY_INSERT is set to OFF.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, DbCommandMethod executeMethod, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
   --- End of inner exception stack trace ---
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(DbContext _, ValueTuple`2 parameters)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(IEnumerable`1 commandBatches, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IReadOnlyList`1 entriesToSave)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
```
So using the non-generic version of `.Add` fixes the duplication issue, but obviously removes the extremely helpful use of generics, without which we would have to rewrite a significant amount of code to not use the generic methods.

I've included a version of the generic method we are using below, with names simplified to avoid confusion and some extra processing removed as it is not relevant to this.

```c#
public void Merge<T>(T item, Guid personId, Expression<Func<T, bool>> condition = null) where T : class, IGuid
        {
            condition = condition ?? (c => true);

            // Get the original row in the database.
            T original = DbContext.Set<T>().AsNoTracking().SingleOrDefault(condition);

            // If there is no row found or the new row doesn't equal the old row.
            if (original == null || item.RowId != original.RowId)
            {
                // This is broken for concurrent requests.
                DbContext.Set<T>().Add(item);
               // DbContext.Accounts.Add(item as Account); fixes the duplication but obviously breaks benefits of using generics.
            }
            // If there is a row that matches the row that's been sent.
            else if (item.RowId == original.RowId && original.PersonId == personId)
            {
                // And Update the database row.
                DbContext.Set<T>().Update(item);
            }
            // If the row no longer exists in the request, we remove it.
            else if (original.RowId != item.RowId && original.PersonId == personId)
            {
                DbContext.Set<T>().Remove(original);
            }
        }
```

### Further technical details
EF Core version: 2.2.6
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 Enterprise Version 1903
IDE: Visual Studio Enterprise 2019 16.1.6

**Comment by ajcvickers** [comment]
@JackSteel97 Couple of things:
* `DbSet<TEntity>.Add()` should behave in exactly the same way as `DbContext.Add<TEntity>()` and `DbContext.Add()`. If you're seeing differences, then we will likely need a small, runnable project/solution or complete code listing that demonstrates this behavior so that we can investigate fully.
* You mention concurrent requests. `DbContext` is not thread safe--the same instance cannot safely be used from concurrent threads. You might want to confirm that this isn't happening.

**Comment by JackSteel97** [comment]
@ajcvickers Thanks for the reply.

- I'll try and get a repro demo sorted in the next couple days.
- I should have clarified, the concurrent requests are to separate instances of the context. The context is dependency injected through the constructor of the controller.

**Comment by JackSteel97** [comment]
After further investigation, while trying to reproduce the issue. It appears that it is not an issue with Entity Framework. It is just a race condition between the item exists check and the add new item statement. It was a coincidence that the non-generic function did not create duplicates the five times I tested it. I've now been able to confirm the non-generic `.Add` will produce the same duplicating behaviour on random occasions. We've been able to fix this by enforcing a unique index on the SQL database side.

From our testing it looks like the generic function causes this duplication to occur more often, we speculated that this may be due to the overhead of using a generic altering the timings in just such a way to make the duplication more likely but can't be sure.
